feat: Add the option to use sse as the protocol for subgraph subscripion

bakjos opened the pull request:
This PR adds support for using graphql sse with the graphql subscriptions. The following config 

```yaml
subscription:
  enabled: true
  mode:
    experimental_sse:
      all:
        path: /stream
      subgraphs:
        rng:
          path: /stream
          backoff_factor: 4
```

This PR also uses a modified version of the [rust-eventsource-client](https://github.com/launchdarkly/rust-eventsource-client) to use `pin-project-lite` instead of `pin-project` 

<!-- start metadata -->

**Checklist**

Complete the checklist (and note appropriate exceptions) before a final PR is raised.

- [x] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [x] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]. It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]. Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]. Tick whichever testing boxes are applicable. If you are adding Manual Tests:
    - please document the manual testing (extensively) in the Exceptions.
    - please raise a separate issue to automate the test and label it (or ask for it to be labeled) as `manual test`

apollo-cla commented:
@bakjos: Thank you for submitting a pull request! Before we can merge it, you'll need to sign the Apollo Contributor License Agreement here: https://contribute.apollographql.com/

bnjjj commented on apollo-router/src/protocols/sse/client.rs:
```diff
@@ -0,0 +1,638 @@
+                },
+                StateProj::WaitingToReconnect { sleep: delay } => {
+                    ready!(delay.poll(cx));
+                    tracing::info!("Reconnecting");
```
```suggestion
                    tracing::info!("Reconnecting sse connection", url = ?self.current_url);
```
I don't know if it compiles but you got the main idea

bnjjj commented on apollo-router/src/protocols/sse/client.rs:
```diff
@@ -0,0 +1,638 @@
+}
+
+fn delay(dur: Duration, description: &str) -> Sleep {
+    tracing::info!("Waiting {:?} before {}", dur, description);
```
```suggestion
    tracing::debug!("Waiting {:?} before {}", dur, description);
```

bnjjj commented on apollo-router/src/protocols/sse/event_parser.rs:
```diff
@@ -0,0 +1,512 @@
+                        .event_data
+                        .get_or_insert_with(|| EventData::new().with_id(id.clone()));
+
+                    if key == "event" {
```
Maybe a `match` would be better in that case ?

bnjjj commented on apollo-router/src/protocols/sse/mod.rs:
```diff
@@ -0,0 +1,138 @@
+                            }
+                        }
+                        if let ServerMessage::Ping { .. } = server_message {
+                            // Send pong asynchronously
```
Is it a TODO ?

bakjos replied:
I removed this part, since it was not really used

bnjjj commented on apollo-router/src/protocols/sse/mod.rs:
```diff
@@ -0,0 +1,138 @@
+                            .error(
+                                graphql::Error::builder()
+                                    .message(format!(
+                                        "cannot deserialize websocket server message: {err:?}"
```
bad description

bnjjj commented on apollo-router/src/protocols/sse/mod.rs:
```diff
@@ -0,0 +1,138 @@
+                                    .message(format!(
+                                        "cannot deserialize websocket server message: {err:?}"
+                                    ))
+                                    .extension_code("INVALID_WEBSOCKET_SERVER_MESSAGE_FORMAT")
```
```suggestion
                                    .extension_code("INVALID_SSE_SERVER_MESSAGE_FORMAT")
```

bnjjj commented on apollo-router/src/services/subgraph_service.rs:
```diff
@@ -348,6 +353,18 @@ impl tower::Service<SubgraphRequest> for SubgraphService {
                         );
                     }
+                    Some(SubscriptionMode::Sse(cfg)) => {
+                        // call_websocket for passthrough mode
```
wrong comment

bnjjj commented on apollo-router/src/services/subgraph_service.rs:
```diff
@@ -422,6 +439,133 @@ impl tower::Service<SubgraphRequest> for SubgraphService {
+    let (mut handle_sink, handle_stream) = handle.split();
+
+    tokio::task::spawn(async move {
+        let mut stream = gql_stream.map(Ok::<_, graphql::Error>);
+
+        while let Some(val) = stream.next().await {
+            if let Ok(val) = val {
+                if let Err(err) = handle_sink.send(val).await {
+                    tracing::trace!(
+                        "cannot send the sse stream to the subscription stream: {err:?}"
+                    );
+                    break;
+                }
+            }
+        }
+
+        if let Err(err) = handle_sink.close().await {
+            tracing::trace!("cannot close the websocket stream: {err:?}");
+        }
```
Why don't you use the `forward` method as used in `call_websocket` ?

bakjos replied:
I tried with `forward`, but `forward` will keep waiting when the sink is closed and will only finish when the stream finishes

bnjjj commented on apollo-router/src/services/subgraph_service.rs:
```diff
@@ -914,6 +1058,76 @@ fn get_websocket_request(
+                status_code: None,
+            })?;
+    }
+    builder = builder.header("Accept", "text/event-stream").map_err(|e| {
```
Could you use the constant https://docs.rs/http/latest/http/header/constant.ACCEPT.html ?

bnjjj commented on apollo-router/src/services/subgraph_service.rs:
```diff
@@ -914,6 +1058,76 @@ fn get_websocket_request(
+            status_code: None,
+        }
+    })?;
+    builder = builder.method("POST".to_string());
```
Same here, use the enum method coming from http crate

bnjjj commented on apollo-router/src/protocols/sse/client.rs:
```diff
@@ -0,0 +1,638 @@
+
+/// Client is the Server-Sent-Events interface.
+/// This trait is sealed and cannot be implemented for types outside this crate.
+pub trait Client: Send + Sync + private::Sealed {
```
Why do we need this trait ? Isn't it enough to just return the concrete struct in the implementation ?

bakjos replied:
To have the option of creating a different client later (not using hyper for instance)

bnjjj commented on apollo-router/src/protocols/sse/client.rs:
Too much `pub` in this file

bnjjj reviewed (commented):
That looks pretty good, just few comments, I saw a lot of `struct` and `enum` exposed as `pub`, could you just expose them in `pub(crate)` it's only used internally so should not be exposed.

I'll have to test it locally with some E2E examples. I also suggest to rename `see` mode in configuration into `experimental_sse` to let us test it correctly before releasing it to stable. Also, it might be interesting to double check whenever the SSE connection is ended from the client or the router side that everything is cleaned up correctly (subscription to the client is cut if it's not deduplicated, all channels and tasks corresponding to this sub are closed too, ...)

bnjjj commented on apollo-router/src/configuration/snapshots/apollo_router__configuration__tests__schema_generation.snap:
```diff
@@ -1773,6 +1775,110 @@ expression: "&schema"
               "nullable": true
+            },
+            "sse": {
+              "description": "Enable passthrough mode for subgraph(s)",
```
this is incorrect

bnjjj commented on apollo-router/src/plugins/subscription.rs:
```diff
@@ -181,6 +212,43 @@ pub(crate) struct WebSocketConfiguration {
 
+#[derive(Debug, Clone, PartialEq, Eq, Default, Deserialize, Serialize, JsonSchema)]
+#[serde(deny_unknown_fields, default)]
+pub(crate) struct SseConfiguration {
+    /// Path use to get the SSE stream
+    pub(crate) path: Option<String>,
+
+    /// If `reconnect` is `true` (the [default]), the client will automatically
+    /// try to reconnect if the stream ends due to an error. If it is `false`,
+    /// the client will stop receiving events after an error.
+    pub(crate) reconnect: Option<bool>,
+
+    /// If `true` (the [default]) the client will automatically retry the connection, with the
+    /// same delay and backoff behaviour as for reconnects due to stream error.
+    /// If `false`, the client will not retry the initial
+    /// connection.
+    pub(crate) retry_initial: Option<bool>,
+
+    /// After an error, the client will wait this long before the first attempt
+    /// to reconnect.  Subsequent reconnect attempts may wait longer, depending
+    /// on the [`backoff_factor`].
+    #[serde(with = "humantime_serde")]
+    #[schemars(with = "Option<String>")]
+    pub(crate) delay: Option<std::time::Duration>,
+
+    /// Configure the factor by which delays between reconnect attempts will
+    /// exponentially increase, up to [`delay_max`]. The [default] factor is 2,
+    /// so each reconnect attempt will wait twice as long as the previous one.
+    pub(crate) backoff_factor: Option<u32>,
+
+    /// Configure the maximum delay between reconnects (the [default] is 1
+    /// minute). The exponential backoff configured by [`backoff_factor`] will
+    /// not cause a delay greater than this value.
+    #[serde(with = "humantime_serde")]
+    #[schemars(with = "Option<String>")]
+    pub(crate) delay_max: Option<std::time::Duration>,
+}
```
As everything is optional, maybe we should add `enabled: bool`  which defaults to `true`. Because in case we just want to enable it with default values it will be an empty object and in yaml it's pretty unusual. What do you think ?

bakjos replied:
It's a bit weird to me having `enabled = false` I made the change and when is not enabled will use the defaults regardless of the values

bnjjj replied:
having a `enabled: false` is also a good way to quickly disable it without deleting all the configuration. But yes I understand your concern. It's the pattern we adopted right now when we are in that kind of situation

bnjjj commented on apollo-router/src/protocols/sse/client.rs:
```diff
@@ -0,0 +1,640 @@
+
+/// Maximum amount of redirects that the client will follow before
+/// giving up, if not overridden via [ClientBuilder::redirect_limit].
+pub const DEFAULT_REDIRECT_LIMIT: u32 = 16;
```
```suggestion
pub(crate) const DEFAULT_REDIRECT_LIMIT: u32 = 16;
```

does it even need to be exported ?

bakjos replied:
Yup, it's not exported anymore

bnjjj commented on apollo-router/src/protocols/sse/client.rs:
```diff
@@ -0,0 +1,640 @@
+use super::retry::BackoffRetry;
+use super::retry::RetryStrategy;
+
+pub type HttpsConnector = RustlsConnector<HttpConnector>;
```
```suggestion
pub(crate) type HttpsConnector = RustlsConnector<HttpConnector>;
```

bnjjj commented on apollo-router/src/protocols/sse/client.rs:
```diff
@@ -0,0 +1,640 @@
+            .map_err(|e| Error::InvalidParameter(Box::new(e)))?;
+
+        let mut header_map = HeaderMap::new();
+        header_map.insert("Accept", HeaderValue::from_static("text/event-stream"));
+        header_map.insert("Cache-Control", HeaderValue::from_static("no-cache"));
```
Use constant from http crate for Accept and Cache-Control

bnjjj commented on apollo-router/src/protocols/sse/client.rs:
```diff
@@ -0,0 +1,640 @@
+}
+
+impl State {
+    fn name(&self) -> &'static str {
```
Should we call this method `as_str()` to follow the Rust convention ? Or is it more accurate to keep `name` ? Just asking I don't have strong opinion

bakjos replied:
I changed it to `as_str`

bnjjj commented on apollo-router/src/protocols/sse/error.rs:
```diff
@@ -0,0 +1,77 @@
+
+/// Error type returned from this library's functions.
+#[derive(Debug)]
+pub enum Error {
```
```suggestion
pub(crate) enum Error {
```

bnjjj commented on apollo-router/src/protocols/sse/error.rs:
```diff
@@ -0,0 +1,77 @@
+}
+
+impl Error {
+    pub fn is_http_stream_error(&self) -> bool {
```
```suggestion
    pub(crate) fn is_http_stream_error(&self) -> bool {
```

bnjjj commented on apollo-router/src/protocols/sse/error.rs:
```diff
@@ -0,0 +1,77 @@
+        false
+    }
+
+    pub fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
```
```suggestion
    pub(crate) fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
```

bnjjj commented on apollo-router/src/protocols/sse/event_parser.rs:
```diff
@@ -0,0 +1,518 @@
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Event {
+    pub event_type: String,
+    pub data: String,
+    pub id: Option<String>,
+    pub retry: Option<u64>,
+}
```
```suggestion
pub(crate) struct Event {
    pub(crate) event_type: String,
    pub(crate) data: String,
    pub(crate) id: Option<String>,
    pub(crate) retry: Option<u64>,
}
```

bnjjj commented on apollo-router/src/protocols/sse/event_parser.rs:
```diff
@@ -0,0 +1,518 @@
+}
+
+#[must_use = "streams do nothing unless polled"]
+pub struct EventParser {
```
```suggestion
pub(crate) struct EventParser {
```

bnjjj commented on apollo-router/src/protocols/sse/event_parser.rs:
```diff
@@ -0,0 +1,518 @@
+}
+
+impl EventParser {
+    pub fn new() -> Self {
```
```suggestion
    pub(crate) fn new() -> Self {
```

bnjjj commented on apollo-router/src/protocols/sse/event_parser.rs:
```diff
@@ -0,0 +1,518 @@
+        }
+    }
+
+    pub fn was_processing(&self) -> bool {
```
```suggestion
    pub(crate) fn was_processing(&self) -> bool {
```

bnjjj commented on apollo-router/src/protocols/sse/event_parser.rs:
```diff
@@ -0,0 +1,518 @@
+        }
+    }
+
+    pub fn get_event(&mut self) -> Option<Sse> {
```
```suggestion
    pub(crate) fn get_event(&mut self) -> Option<Sse> {
```

bnjjj commented on apollo-router/src/protocols/sse/event_parser.rs:
```diff
@@ -0,0 +1,518 @@
+        self.sse.pop_front()
+    }
+
+    pub fn process_bytes(&mut self, bytes: Bytes) -> Result<()> {
```
```suggestion
    pub(crate) fn process_bytes(&mut self, bytes: Bytes) -> Result<()> {
```

bnjjj commented on apollo-router/src/protocols/sse/client.rs:
```diff
@@ -0,0 +1,640 @@
+use hyper::body::HttpBody;
+use hyper::client::connect::Connect;
+use hyper::client::connect::Connection;
+pub use hyper::client::HttpConnector;
```
Should it be `pub`?

bnjjj commented on apollo-router/src/protocols/sse/client.rs:
```diff
@@ -0,0 +1,640 @@
+
+pin_project! {
+    #[must_use = "streams do nothing unless polled"]
+    pub struct ReconnectingRequest<C> {
```
```suggestion
    pub(crate) struct ReconnectingRequest<C> {
```

bnjjj commented on apollo-router/src/protocols/sse/mod.rs:
```diff
@@ -0,0 +1,135 @@
+use crate::graphql;
+use crate::protocols::websocket::ServerError;
+
+pub mod client;
```
```suggestion
pub(crate) mod client;
```

bnjjj reviewed (commented):
This PR should not contain any `pub` I think

bakjos commented on apollo-router/src/protocols/sse/client.rs:
```diff
@@ -0,0 +1,640 @@
+
+/// Client is the Server-Sent-Events interface.
+/// This trait is sealed and cannot be implemented for types outside this crate.
+pub(crate) trait Client: Send + Sync + private::Sealed {
```
I changed it to use the concrete type

bakjos commented:
@bnjjj I changed all the methods/types and structs to be internal

bnjjj commented:
@bakjos thanks, it's time for me to test it locally, could you tell me what kind of subgraph are you using ? If you have a public repo I can clone to test it that would be great, if not I'll have to build it by myself :)

bakjos commented:
I don't have a public repo I can share, however for some tests I was using this one on deno but it doesn't have any key attributes
```typescript

import { serve } from 'https://deno.land/std/http/server.ts';
import { createHandler } from 'npm:graphql-sse/lib/use/fetch';
import { GraphQLSchema, GraphQLObjectType, GraphQLString } from 'npm:graphql';

console.log('Creating Schema...');
const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

/**
 * Construct a GraphQL schema and define the necessary resolvers.
 *
 * type Query {
 *   hello: String
 * }
 * type Subscription {
 *   greetings: String
 * }
 */
const schema = new GraphQLSchema({
  query: new GraphQLObjectType({
    name: 'Query',
    fields: {
      hello: {
        type: GraphQLString,
        resolve: () => 'world',
      },
    },
  }),
  subscription: new GraphQLObjectType({
    name: 'Subscription',
    fields: {
      greetings: {
        type: GraphQLString,
        subscribe: async function* () {
          for (const hi of ['Hi', 'Bonjour', 'Hola', 'Ciao', 'Zdravo']) {
            await sleep(1000);
            yield { greetings: hi };
          }
        },
      },
    },
  }),
});

// Create the GraphQL over SSE native fetch handler
const handler = createHandler({ schema });

let port = 4000;
if (Deno.args.length > 0) {
  port = parseInt(Deno.args[0]);
}

console.log('Starting server...');
// Serve on `/graphql/stream` using the handler
await serve(
  async (req: Request) => {
  
    const [path, _search] = req.url.split('?');
    
    if (path.endsWith('/graphql/stream')) {
      console.log('executing handler', req);
      return await handler(req);
    }
    return new Response(null, { status: 404 });
  },
  {
    port, // Listening to port 4000
  },
);

```

bnjjj commented:
@bakjos I'm not able to reproduce, I have an error `{"payload":null,"errors":[{"message":"cannot read message from sse: UnexpectedResponse(415)","extensions":{"code":"SSE_MESSAGE_ERROR"}}]}` maybe I'm missing some configuration ?

Here is mine:

```yaml
mode:
    experimental_sse:
      all:
        path: /graphql/stream
```

bnjjj commented:
Ok I was able to test it locally @bakjos , few things I already detected. If you open a subscription and stop the subgraph server delivering the subscription it's trying to reconnect forever. By default we don't want to reconnect on subscription, it's better to reconnect from the client by default because we know we might loose some events. Also even if after that I'm closing the subscription from the client side, I still see the logs telling me it's trying to reconnect to the sse stream but even if it's able to reconnect nothing is still plugged to that stream so it should directly be stopped and shouldn't have a task running in the background for nothing.

bakjos commented:
Thanks @bnjjj for checking it locally, I did tested it when the subscription was closed from the client, but the subgraph was still running, I couldn't find another way besides adding a closed flag using `AtomicBool` to check if the connection was closed and finish the loop, I also changed it to not reconnect by default

bnjjj commented:
Thanks @bakjos but I think we can do something better than just having an interval stream responsible to check if the connection is closed. Let me explain how it works for websocket for example, which is pretty similar to how it should work for sse too I think. When I'm receiving a `Handle` from `notification.rs` I can split them as a `Stream` and a `Sink`. If the `Sink` is dropped and we don't have any other existing `Sink` then automatically when you're trying to listen on the `Stream` it will return a `None` which means the `Stream` is closed. I think somewhere (I didn't dig in details) you're keeping a `Sink` in a task. In fact as soon as we received an error from the original SSE stream or when it's disconnected we should directly close the task and the `Sink`, by doing this if you're cleaning correctly everything it should just clean all the channels and topic automatically in `notification.rs`. If you want more information or more real time interaction feel free to go on discord I sent you messages there :) Happy to help you to debug this

bakjos commented:
@bnjjj I added the timeout and switched to using a channel instead of a timeout, what's what you think is wrong with adding the channel to the Handle?

bnjjj commented:
Hello @bakjos I just opened a new PR with some changes https://github.com/bakjos/router/pull/1 feel free to merge it if you agree (do not forget to merge with our current dev branch to avoid noises in that PR). Also I tried to trigger to timeout on the reconnect and it didn't work. Could take a look please ?

bakjos commented:
@bnjjj Thank you so much for adding those changes, will test the timeout again, do you have the configuration you used to test the timeout on reconnect?

bnjjj commented:
@bakjos BTW if you can avoid force-pushing that would help me a lot to review only changes you made since my last review. At the end I will squash all your commits so that's not an issue. 

I used the default value for timeout which is now `60s`, I didn't check if it was really set correctly so it might be the root cause. Anyway, here is the configuration I set locally:

```yaml
mode:
    experimental_sse:
      all:
        path: /graphql/stream
        reconnect: true
        delay_max: 15s
```

bakjos commented:
@bnjjj I pushed a new commit, The reason why it was not working was that it was taking 1800s instead of 60s as the default value, I reverted some of the changes because it was not breaking the loop when an error was found, so the connection was not terminated on a Reconnect timeout

bnjjj commented on apollo-router/src/plugins/subscription.rs:
```diff
@@ -182,6 +213,96 @@ pub(crate) struct WebSocketConfiguration {
+    /// If `reconnect` is `true`, the client will automatically
+    /// try to reconnect if the stream ends due to an error. If it is `false`  (the [default]),
+    /// the client will stop receiving events after an error.
+    pub(crate) reconnect: Option<bool>,
```
As we're using `.unwrap_or_default()` when using for `ReconnectOptions` I suggest to change it to `bool` instead of Option. It will still work because of the serde default.

bnjjj commented on apollo-router/src/plugins/subscription.rs:
```diff
@@ -182,6 +213,96 @@ pub(crate) struct WebSocketConfiguration {
+    /// same delay and backoff behaviour as for reconnects due to stream error.
+    /// If `false` (the [default]), the client will not retry the initial
+    /// connection.
+    pub(crate) retry_initial: Option<bool>,
```
Same here for the Option

bakjos replied:
Done

bnjjj reviewed (commented):
Some comments. It's looking good. I'll try to test it locally ASAP

bakjos commented:
@bnjjj  I updated the branch to the latest dev, should I do something else?

hwillson commented:
Thanks very much for submitting this PR @bakjos, and sorry for the delay in responding! Your PR has led to quite a bit of discussion internally. 😅 There are 2 main reasons why we haven't merged this PR in:

1. You are submitting a PR against commercial functionality in the Router. While we definitely appreciate this, we just want to be clear that users of your submitted changes will have to have a commercial license to use them. Knowing this, would you still like to submit these changes?
2. Adding SSE support to the Router is great way to unblock the migration of existing SSE enabled graphs to Federation. While this is definitely appealing, it's important to know what we're going to be introducing [multipart subscriptions](https://www.apollographql.com/docs/router/executing-operations/subscription-multipart-protocol) support, between the Router and subgraphs, in the not too distant future. This will be our preferred and recommended (passthrough mode) way of using subscriptions between the Router and subgraphs. We are a bit hesitant to add SSE support now knowing it is not part of our recommended subscriptions workflow.

Let us know how you feel about 1 - we're still deciding on 2, but we'll need an answer to 1 before continuing regardless. Thanks!

bakjos commented:
@hwillson Thanks for the response it makes sense, I'm fine with number 1

abernix commented:
Hi there.  I'll have to close this PR for now since the core team is not committed or aligned to supporting SSE between Router and subgraphs at this time.   It may become an option we're willing to consider and carry in the future, but for now, we're just not able to commit to prioritizing the carrying cost and maintenance involved with bringing SSE along (and everything that comes with it, benchmarking, documentation, etc.)

We can of course re-open this if the time comes, and we do believe that time could come, but it would have to be at another time of in the project.  Thanks for making the contribution attempt originally.
